### PR TITLE
azimuth fm rate mismatch -> lut.py

### PR DESCRIPTION
**Pull request by seongsujeong**
This PR is to add the azimuth FM rate mismatch into the model-based LUTs i.e. `lut.py`

~~- `compute_geocoding_correction_luts` will skip the azimuth FM rate mismatch when DEM is not provided.~~ After the discussion, its behavior was changed to return error when DEM is not provided.
~~- If it is calculated, it will be added to the azimuth component of the return value.~~ The azimuth FM rate mismatch will be returned separately, so that it can be written to the output separately.

### CODE REVIEW

**Review by LiangJYu — DISMISSED** [review]
LGTM

File: src/compass/utils/lut.py [hunk 1]
```diff
@@ -11,6 +11,13 @@ def compute_geocoding_correction_luts(burst, rg_step=200, az_step=0.25):
         LUT spacing (in pixels) along x/slant range in meters
     ystep: int
         LUT spacing (in pixels) along x/azimuth in seconds
```
Comment (hfattahi) [inline src/compass/utils/lut.py:13]: fix the naming and units

File: src/compass/utils/lut.py [hunk 2]
```diff
@@ -11,6 +11,13 @@ def compute_geocoding_correction_luts(burst, rg_step=200, az_step=0.25):
...
+    path_dem: str
+        Path to the DEM required for azimuth FM rate mismatch.
+        If `None`, the calculation of the FM mismatch will be skipped.
+    path_scratch: str
```
Comment (yunjunz) [inline src/compass/utils/lut.py:17]: I would suggest using `dem_path` and `scratch_path` in style, to be more consistent with existing ones: https://github.com/opera-adt/s1-reader/blob/ae88a1d56df19b8e1cd5f8be9bc73e195e256c31/src/s1reader/s1_reader.py#L472-L473

File: src/compass/utils/lut.py [hunk 3]
```diff
@@ -23,7 +30,19 @@ def compute_geocoding_correction_luts(burst, rg_step=200, az_step=0.25):
     az_bistatic = burst.bistatic_delay(range_step=rg_step, az_step=az_step)
     rg_doppler = burst.geometrical_and_steering_doppler(range_step=rg_step, az_step=az_step)
 
+    if path_dem is None:
+        # Skip the FM mismatch rate calculation
+        az_fm_mismatch = None
```
Comment (hfattahi) [inline src/compass/utils/lut.py:35]: I think the check should just check if the DEM is valid. If DEM is not valid then raise an exception.

File: src/compass/utils/lut.py [hunk 4]
```diff
@@ -23,7 +30,19 @@ def compute_geocoding_correction_luts(burst, rg_step=200, az_step=0.25):
     az_bistatic = burst.bistatic_delay(range_step=rg_step, az_step=az_step)
     rg_doppler = burst.geometrical_and_steering_doppler(range_step=rg_step, az_step=az_step)
```
Comment (hfattahi) [inline src/compass/utils/lut.py:31]:
```suggestion
    rg_lut = burst.geometrical_and_steering_doppler(range_step=rg_step, az_step=az_step)
```

File: src/compass/utils/lut.py [hunk 5]
```diff
@@ -23,7 +30,19 @@ def compute_geocoding_correction_luts(burst, rg_step=200, az_step=0.25):
...
+    if path_dem is None:
+        # Skip the FM mismatch rate calculation
+        az_fm_mismatch = None
+    else:
+        az_fm_mismatch = burst.az_fm_rate_mismatch_mitigation(path_dem,
+                                                              path_scratch,
+                                                              rg_step=rg_step,
+                                                              az_step=az_step)
+
     rg_lut = rg_doppler
```
Comment (hfattahi) [inline src/compass/utils/lut.py:42]: not needed

File: src/compass/utils/lut.py [hunk 6]
```diff
@@ -23,7 +30,19 @@ def compute_geocoding_correction_luts(burst, rg_step=200, az_step=0.25):
...
     az_lut = az_bistatic.data
 
-    return rg_lut, az_lut
\ No newline at end of file
+    if not az_fm_mismatch is None:
+        az_lut += az_fm_mismatch.data
```
Comment (hfattahi) [inline src/compass/utils/lut.py:46]: let's coordinate with @yunjunz and make sure the signs are consistent with ETAD and we can add the two.

**Review by hfattahi — COMMENTED** [review]
Thanks @seongsujeong . Some comments here

File: src/compass/utils/lut.py (hunk 6 above)
Comment (LiangJYu) [inline src/compass/utils/lut.py:46]: The types don't match to perform the `+=`. I think the following will get the correct result.
```suggestion
        az_lut_data = az_bistatic.data + az_fm_mismatch.data
        az_lut = isce3.core.LUT2d(az_bistatic.x_start, az_bistatic.y_start, 
                az_bistatic.x_spacing, az_bistatic.x_spacing, az_lut_data)
```

File: src/compass/utils/lut.py [hunk 7]
```diff
@@ -23,7 +30,19 @@ def compute_geocoding_correction_luts(burst, rg_step=200, az_step=0.25):
...
 
+    if path_dem is None:
+        # Skip the FM mismatch rate calculation
+        az_fm_mismatch = None
+    else:
+        az_fm_mismatch = burst.az_fm_rate_mismatch_mitigation(path_dem,
+                                                              path_scratch,
+                                                              rg_step=rg_step,
+                                                              az_step=az_step)
+
```
Comment (LiangJYu) [inline src/compass/utils/lut.py:41]: I don't like the redundant rename of the variables below. Can we do the following to at the beginning of this function:
```python
    az_lut = burst.bistatic_delay(range_step=rg_step, az_step=az_step)
    rg_lut = burst.geometrical_and_steering_doppler(range_step=rg_step, az_step=az_step)
```

**Review by LiangJYu — COMMENTED** [review]
Some of my comments aren't for changes in this PR but would be nice to have since you're working on this right now.

File: src/compass/utils/lut.py (hunk 6 above)
Comment (seongsujeong) [reply src/compass/utils/lut.py]: @yunjunz The difference between our calculation and the ETAD data was small. So the sign of our calculation and the ETAD data would be the same. Therefore, I think the sign to apply our azimuth FM rate mismatch mitigation has to be the same as we do with ETAD data.

Wound you be able to confirm if we are "adding" the correction when we are using ETAD?

File: src/compass/utils/lut.py [hunk 8]
```diff
@@ -1,4 +1,12 @@
-def compute_geocoding_correction_luts(burst, rg_step=200, az_step=0.25):
+'''
+Placeholder for model-based correction LUT
+'''
+import os
+
```
Comment (vbrancat) [inline src/compass/utils/lut.py:5]:
```suggestion

```
Remove space

File: src/compass/utils/lut.py [hunk 9]
```diff
@@ -8,22 +16,44 @@ def compute_geocoding_correction_luts(burst, rg_step=200, az_step=0.25):
...
         2D array containg sum of azimuth corrections
     '''
 
-    az_bistatic = burst.bistatic_delay(range_step=rg_step, az_step=az_step)
-    rg_doppler = burst.geometrical_and_steering_doppler(range_step=rg_step, az_step=az_step)
+    az_lut = burst.bistatic_delay(range_step=rg_step, az_step=az_step)
+    rg_lut = burst.geometrical_and_steering_doppler(range_step=rg_step, az_step=az_step)
+
+    if dem_path is None:
+        raise ValueError('DEM for azimith FM rate mismatch was not provided.')
```
Comment (vbrancat) [inline src/compass/utils/lut.py:42]:
```suggestion
        raise ValueError('DEM for azimuth FM rate mismatch was not provided.')
```

File: src/compass/utils/lut.py (hunk 6 above)
Comment (vbrancat) [reply src/compass/utils/lut.py]: Please @yunjunz, confirm what @seongsujeong has mentioned above. Also to my understanding, to be consistent with ETAD, we should add the two corrections

**Review by vbrancat — COMMENTED** [review]
@seongsujeong LGTM, I will wait for @yunjunz to confirm the sign of the correction to be consistent with ETAD) and then I will approve the PR. Thanks :)

**Comment by vbrancat** [comment]
Just a note. If I am not totally out of sync, I think we agreed to save the model-based corrections within the output CSLC HDF5 file. @LiangJYu is crafting a PR to save the bistatic delay and the Doppler within the product. Hopefully, this will be merged into `main `ASAP. If we are still shooting to include the correction within the product (let's discuss about it tomorrow), make sure that also the azimuth FM-rate mismatch gets saved.

**Review by LiangJYu — APPROVED** [review]
LGTM

File: src/compass/utils/lut.py [hunk 10]
```diff
@@ -1,4 +1,12 @@
-def compute_geocoding_correction_luts(burst, rg_step=200, az_step=0.25):
+'''
+Placeholder for model-based correction LUT
+'''
+import os
+
+import isce3
+
+def compute_geocoding_correction_luts(burst, rg_step=200, az_step=0.25,
+                                      dem_path=None, scratch_path=None):
```
Comment (LiangJYu) [inline src/compass/utils/lut.py:9]:
```suggestion
                                      dem_path, scratch_path=None):
```
Let's remove default value if this is a required parameter.

Comment (yunjunz) [reply src/compass/utils/lut.py]: I agree with this; and move the required `dem_path` arg location to the 2nd place.

File: src/compass/s1_geocode_slc.py [hunk 11]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env python
 
-'''wrapper for geocoded SLC'''
+'''wrapper for geocoded CSLC'''
```
Comment (yunjunz) [inline src/compass/s1_geocode_slc.py:3]:
```suggestion
'''wrapper for geocoded SLC'''
```

Geocoded SLC is more commonly used and accepted term, I would suggest keeping the old name.

File: src/compass/utils/lut.py (hunk 6 above)
Comment (yunjunz) [reply src/compass/utils/lut.py]: My testings with `ISCE2/topsApp.py` shows the `geometrical offset - ETAD offset` gives the correct result (fewer phase artifacts due to SAR processing effects), thus, if our `az_lut` here is consistent with isce2, the sign should be `-=`.

Comment (yunjunz) [reply src/compass/utils/lut.py]: Regarding https://github.com/opera-adt/COMPASS/pull/72#discussion_r1047826363, yes, we should include all 3 components in the final offsets.

File: src/compass/s1_geocode_slc.py (hunk 11 above)
Comment (LiangJYu) [reply src/compass/s1_geocode_slc.py:3]: In discussions in #75, @vbrancat suggested [`CSLC`](https://github.com/opera-adt/COMPASS/blob/a5138d09e37c60223b1ed6259d652d8bdacf15ad/src/compass/s1_geocode_slc.py#L137).

Comment (yunjunz) [reply src/compass/s1_geocode_slc.py:3]: I was not aware of it. Thanks @LiangJYu

**Comment by seongsujeong** [comment]
FYI - There will be  following commits that is in line with the PR #75. The highlights of the changes are:
- `lut.compute_geocoding_correction_luts()` will return three LUTs (bistatic, total doppler, azimuth FM rate mismatch), rather than azimuth / range components of the correction
- Change the function call in `s1_geocode_slc.py` accordingly;
- Write azimuth FM rate mismatch into output `.h5`, group `/science/SENTINEL1/CSLC/corrections`

**Review by vbrancat — APPROVED** [review]
Btw, I was testing the PR and I found this issue:

```
(compass) [vbrancat@aurora COMPASS]$ python src/compass/s1_cslc.py /mnt/aurora-r0/vbrancat/data/S1/runconfig/runconfigs/geo_cslc_s1.yaml 
Traceback (most recent call last):
  File "/mnt/aurora-r0/vbrancat/codes/COMPASS/src/compass/s1_cslc.py", line 5, in <module>
    from compass import s1_rdr2geo, s1_geo2rdr, s1_resample, s1_geocode_slc
  File "/mnt/aurora-r0/vbrancat/codes/miniconda3/envs/compass/lib/python3.11/site-packages/compass/s1_geocode_slc.py", line 18, in <module>
    from compass.utils.h5_helpers import (corrections_to_h5group,
  File "/mnt/aurora-r0/vbrancat/codes/miniconda3/envs/compass/lib/python3.11/site-packages/compass/utils/h5_helpers.py", line 15, in <module>
    from compass.utils.lut import compute_geocoding_correction_luts
  File "/mnt/aurora-r0/vbrancat/codes/miniconda3/envs/compass/lib/python3.11/site-packages/compass/utils/lut.py", line 8
    dem_path, scratch_path=None):
    ^^^^^^^^
SyntaxError: non-default argument follows default argument
not sure where it comes from
```

**Review by vbrancat — COMMENTED** [review]
Btw, I was testing the PR and I found this issue:

```
(compass) [vbrancat@aurora COMPASS]$ python src/compass/s1_cslc.py /mnt/aurora-r0/vbrancat/data/S1/runconfig/runconfigs/geo_cslc_s1.yaml 
Traceback (most recent call last):
  File "/mnt/aurora-r0/vbrancat/codes/COMPASS/src/compass/s1_cslc.py", line 5, in <module>
    from compass import s1_rdr2geo, s1_geo2rdr, s1_resample, s1_geocode_slc
  File "/mnt/aurora-r0/vbrancat/codes/miniconda3/envs/compass/lib/python3.11/site-packages/compass/s1_geocode_slc.py", line 18, in <module>
    from compass.utils.h5_helpers import (corrections_to_h5group,
  File "/mnt/aurora-r0/vbrancat/codes/miniconda3/envs/compass/lib/python3.11/site-packages/compass/utils/h5_helpers.py", line 15, in <module>
    from compass.utils.lut import compute_geocoding_correction_luts
  File "/mnt/aurora-r0/vbrancat/codes/miniconda3/envs/compass/lib/python3.11/site-packages/compass/utils/lut.py", line 8
    dem_path, scratch_path=None):
    ^^^^^^^^
SyntaxError: non-default argument follows default argument
```

not sure where it comes from

**Comment by seongsujeong** [comment]
@vbrancat 

> Btw, I was testing the PR and I found this issue:
> 
> ```
> (compass) [vbrancat@aurora COMPASS]$ python src/compass/s1_cslc.py /mnt/aurora-r0/vbrancat/data/S1/runconfig/runconfigs/geo_cslc_s1.yaml 
> Traceback (most recent call last):
>   File "/mnt/aurora-r0/vbrancat/codes/COMPASS/src/compass/s1_cslc.py", line 5, in <module>
>     from compass import s1_rdr2geo, s1_geo2rdr, s1_resample, s1_geocode_slc
>   File "/mnt/aurora-r0/vbrancat/codes/miniconda3/envs/compass/lib/python3.11/site-packages/compass/s1_geocode_slc.py", line 18, in <module>
>     from compass.utils.h5_helpers import (corrections_to_h5group,
>   File "/mnt/aurora-r0/vbrancat/codes/miniconda3/envs/compass/lib/python3.11/site-packages/compass/utils/h5_helpers.py", line 15, in <module>
>     from compass.utils.lut import compute_geocoding_correction_luts
>   File "/mnt/aurora-r0/vbrancat/codes/miniconda3/envs/compass/lib/python3.11/site-packages/compass/utils/lut.py", line 8
>     dem_path, scratch_path=None):
>     ^^^^^^^^
> SyntaxError: non-default argument follows default argument
> ```
> 
> not sure where it comes from


That was because the default values for `dem_path` were removed, but placed after `range_step` and `rg_step`, which have the default values. The issue was fixed in the following commits.

**Review by vbrancat — APPROVED** [review]
LGTM, thanks :)

**Pull request merged**